meta.go

```diff
@@ -241,16 +242,110 @@ func parseStringList(list interface{}) []string {
 	return result
 }
 
-var termNameRE = regexp.MustCompile("^[a-z]+([a-z0-9-]+)(/[0-9]+)?$")
+var validTermName = regexp.MustCompile(`^[a-zA-Z][\w-]+$`)
 
-func checkTerm(s string) error {
-	match := termNameRE.FindStringSubmatch(s)
-	if match == nil {
-		return fmt.Errorf("invalid term name %q: must match %s", s, termNameRE.String())
+// TermID represents a single term id. The term can either be owned
+// or "public" (meaning there is no owner).
+type TermID struct {
+	Tenant   string
+	Owner    string
+	Name     string
+	Revision int
+}
+
+// Validate returns an error if the Term contains invalid data.
+func (t *TermID) Validate() error {
+	if t.Tenant != "" && t.Tenant != "cs" {
+		if !validTermName.MatchString(t.Tenant) {
+			return fmt.Errorf("wrong term tenant format %q", t.Tenant)
+		}
+	}
+	if t.Owner != "" && !names.IsValidUser(t.Owner) {
+		return fmt.Errorf("wrong owner format %q", t.Owner)
+	}
+	if !validTermName.MatchString(t.Name) {
+		return fmt.Errorf("wrong term name format %q", t.Name)
+	}
+	if t.Revision < 0 {
+		return fmt.Errorf("negative term revision")
 	}
 	return nil
 }
 
+// ParseTerm takes a termID as a string and parses it into a Term.
+// A complete term is in the form:
+// tenant:owner/name/revision
+// This function accepts partially specified identifiers
+// typically in one of the following forms:
+// name
+// owner/name
+// owner/name/27 # Revision 27
+// name/283 # Revision 283
+// cs:owner/name # Tenant cs
+func ParseTerm(s string) (*TermID, error) {
+	tenant := ""
+	termid := s
+	if t := strings.SplitN(s, ":", 2); len(t) == 2 {
+		tenant = t[0]
+		termid = t[1]
+	}
+
+	tokens := strings.Split(termid, "/")
+	var term TermID
+	switch len(tokens) {
+	case 1: // "name"
+		term = TermID{
+			Tenant:   tenant,
+			Owner:    "",
```

Comment:
I'd omit the zero-valued fields. That's part of why the tagged-literal form is nicer.